### PR TITLE
Add asynchronous OV IR inference 

### PR DESCRIPTION
**Pull request by kprokofi**
### Checklist

- [x] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/model/entity/base.py
Comment (sungmanc) [inline src/otx/core/model/entity/base.py]: For me, `use_tp_mode` is unreadable. How about `use_throughput_mode`? It's a little bit long, though.

File: src/otx/core/model/entity/base.py
Comment (sungmanc) [inline src/otx/core/model/entity/base.py:258]: So, it means RGB -> BGR? Do OV_API needs the BGR images?

File: src/otx/core/model/entity/classification.py
Comment (sungmanc) [inline src/otx/core/model/entity/classification.py]: It seems wrong correction since it's classname is `OVMulticlassClassificationModel`

File: src/otx/core/model/entity/classification.py
Comment (sungmanc) [inline src/otx/core/model/entity/classification.py]: I prefer below since it reduces the comment line.
```python3
pred_labels = [torch.tensor(out.top_labels[0][0], dtype=torch.long) for out in outputs]
pred_scores = [torch.tensor(out.top_labels[0][2]) for out in outputs]
```

File: tests/integration/cli/test_cli.py
Comment (sungmanc) [inline tests/integration/cli/test_cli.py:148]: How about maintaining this issue more explicitly ? by using JIRA or Github issue

File: src/otx/core/data/entity/classification.py
Comment (vinnamkim) [inline src/otx/core/data/entity/classification.py:73]: Did you check conflicts properly?

File: src/otx/core/model/entity/base.py
Comment (vinnamkim) [inline src/otx/core/model/entity/base.py:215]: Why don't we just
```suggestion
class OVModel(OTXModel):
```

File: src/otx/core/model/entity/base.py
Comment (vinnamkim) [inline src/otx/core/model/entity/base.py:282]: Is it still synchronous for each batch? I'd rather to make `self.async_inference=True` as the default (batch-wise synchronized inference) for `forward()`. Then, create `def forward_async()` which is truly asynchronous, returning a future of model output. If it is hard to develop `def forward_async()` this time. Please do it later.

File: src/otx/core/utils/build.py
Comment (vinnamkim) [inline src/otx/core/utils/build.py]: Can we notify this behavior to an user such as
```suggestion
    reqs_num = os.cpu_count()
    reqs_num = max(1, int(reqs_num / 2)) if reqs_num is not None else 1
    msg = f"Set the number of OpenVINO inference requests to {reqs_num} because you did not configure 'max_num_requests'"
    warnings.warn(msg)
    return reqs_num
```

File: src/otx/core/model/entity/base.py
Comment (vinnamkim) [reply src/otx/core/model/entity/base.py:258]: It's not about changing color channel. It is about changing an order of tensor dimensions,
`C x H x W => H x W x C`.

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py]: Ok

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py:258]: Correct, to infer OV model, we need to run np.arrays with H x W x C dimension. 
I checked that BGR is by default used in our OTX. Answering to your question, OV requires BGR images, not RGB

File: src/otx/core/model/entity/classification.py
Comment (kprokofi) [reply src/otx/core/model/entity/classification.py]: Oh, you are right, thanks

File: tests/integration/cli/test_cli.py
Comment (kprokofi) [reply tests/integration/cli/test_cli.py:148]: We don't need it. When we add export we change integration tests for OV to e2e regime. It is just to not break inference in the whole. OMZ is not developed anymore, only support current models.

File: src/otx/core/data/entity/classification.py
Comment (kprokofi) [reply src/otx/core/data/entity/classification.py:73]: I checked train and test with torch model. We have same interface for other tasks, I am not sure why we added torch.stack here. @sungmanc do you have comments/concerns here?

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py:215]: I did it at first time, however, it leads to mypy issues, because of the types override mismatch with the base class. Adding Generic type helps. Other solution is to use #noqa

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py:282]: why should we create forward_async()? It works OK in forward() as well. self.async_inference depends on parameter in recipe.yaml. By default it is True for all tasks.

File: src/otx/core/utils/build.py
Comment (kprokofi) [reply src/otx/core/utils/build.py]: We can change it similar to this, yes

File: src/otx/core/data/entity/classification.py
Comment (kprokofi) [reply src/otx/core/data/entity/classification.py:71]: Ok, I see where can be a problem during validation. I will apply fix accordingly

File: src/otx/core/utils/build.py
Comment (kprokofi) [reply src/otx/core/utils/build.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/1da85964b12868bb793297fab408ea395fe88500)

File: src/otx/core/model/entity/classification.py
Comment (kprokofi) [reply src/otx/core/model/entity/classification.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/1da85964b12868bb793297fab408ea395fe88500)

File: src/otx/core/model/entity/classification.py
Comment (kprokofi) [reply src/otx/core/model/entity/classification.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/1da85964b12868bb793297fab408ea395fe88500)

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/1da85964b12868bb793297fab408ea395fe88500)

File: src/otx/core/data/entity/classification.py
Comment (vinnamkim) [reply src/otx/core/data/entity/classification.py:73]: It came from https://github.com/openvinotoolkit/training_extensions/pull/2749

File: src/otx/core/model/entity/base.py
Comment (vinnamkim) [reply src/otx/core/model/entity/base.py:282]: You are doing this
image
But I was asking you change it as a default (implementation) of `forward()` and create this one
image

**Review by jaegukhyun — DISMISSED** [review]

File: src/otx/core/data/entity/classification.py
Comment (sungmanc) [reply src/otx/core/data/entity/classification.py:73]: I'm not sure how it works without error. `OTXModel(nn.Module)` might raise the error since `batch_data.images` is `List[tv_tensors.Image]` while it expects the `Tensor`. After the PR what @vinnamkim pointed out, there is no `torch.stack` in the `_customized_input`.

Comment (kprokofi) [reply src/otx/core/data/entity/classification.py:73]: I will merge and take a look one more time

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py:282]: Let's discuss it later, we now need to focus on export

File: src/otx/core/data/entity/classification.py
Comment (kprokofi) [reply src/otx/core/data/entity/classification.py:73]: OTXBatchDataEntity can receive list of tensors as well. There won't be problems. Training and validation is working
image

Comment (kprokofi) [reply src/otx/core/data/entity/classification.py:73]: We can't use torch stack for OV IR, since it will require resizing operation, but Model_API do it inside. We shouldn't resize it before. Other tasks are also use list[tensor] in batch

Comment (kprokofi) [reply src/otx/core/data/entity/classification.py:73]: Please, let me know if it will raise any issues. For now, it seems to work OK, integration tests prove it

**Review by vinnamkim — DISMISSED** [review]

File: src/otx/core/model/entity/base.py
Comment (sovrasov) [inline src/otx/core/model/entity/base.py:230]: These optional config members look like a workaround. I suggest conducting a refactoring in the future to make it regular

File: src/otx/core/model/entity/base.py
Comment (sovrasov) [reply src/otx/core/model/entity/base.py:258]: Previously, we were using bgr-rgb swap in OTX 1.x, since the original torch models are all trained in RGB, and model API doesn't color channels order.
Does OTX 2.0 train models in BRG?

**Review by sovrasov — DISMISSED** [review]

**Review by jaegukhyun — DISMISSED** [review]
> You are doing this image But I was asking you change it as a default (implementation) of `forward()` and create this one image


It is worth to think about this

File: src/otx/core/model/entity/base.py
Comment (samet-akcay) [inline src/otx/core/model/entity/base.py:228]: Can you explain the motivation why `num_classes` defined as a separate argument, while the rest of the args are defined within the config?

Comment (samet-akcay) [reply src/otx/core/model/entity/base.py:228]: Is it possible to explicitly define these args here. `config` sort of hides them, and it is hard to know what is inside  this `config`

File: src/otx/core/model/entity/base.py
Comment (samet-akcay) [inline src/otx/core/model/entity/base.py]: can you provide more description/context here,
- what is OVModel?
- why do we need it?
- what context we would use the `OVModel`

File: src/otx/core/model/entity/classification.py
Comment (samet-akcay) [inline src/otx/core/model/entity/classification.py:334]: can `outputs` be `Any`? I see that `out` has `top_labels` attribute. can you explain what `outputs` actually contain?

File: src/otx/core/model/entity/detection.py
Comment (samet-akcay) [inline src/otx/core/model/entity/detection.py]:
```suggestion
        for output in outputs:
            output_objects = output.objects
```

File: src/otx/core/model/entity/instance_segmentation.py
Comment (samet-akcay) [inline src/otx/core/model/entity/instance_segmentation.py]:
```suggestion
        for output in outputs:
            output_objects = output.segmentedObjects
```

File: src/otx/core/model/entity/instance_segmentation.py
Comment (samet-akcay) [inline src/otx/core/model/entity/instance_segmentation.py]: In here `src/otx/core/model/entity/segmentation.py` we define the class name as `OVSegmentationModel`. In terms of consistency, can we follow the same convention and use `OVInstanceSegmentationModel`. It might be a bit longer, but looks a lot cleaner and readable

File: src/otx/core/model/module/segmentation.py
Comment (samet-akcay) [inline src/otx/core/model/module/segmentation.py:91]: Not sure if this function is clear. Why is it not possible to pass `preds` and `inputs` to `self.test_metrics.udpate()`? 
I'm aware that it is not the scope of this PR, maybe we could address this in the future.

Comment (samet-akcay) [reply src/otx/core/model/module/segmentation.py:91]: If this conversion is really needed, test_metric should ideally handle this conversion within update. Am I missing anything here?

Comment (samet-akcay) [reply src/otx/core/model/module/segmentation.py:91]: Also is it possible to rename `pred_list` to `predictions` to improve the readability

File: src/otx/core/utils/build.py
Comment (samet-akcay) [inline src/otx/core/utils/build.py]: How about 
```suggestion
def get_default_num_async_infer_requests() -> int:
```
or `get_default_num_async_infer_reqs`?

a bit longer but easier to read

File: src/otx/core/utils/build.py
Comment (samet-akcay) [inline src/otx/core/utils/build.py]: ?
```suggestion
    num_requests = os.cpu_count()
```
or `num_reqs`?

File: src/otx/recipe/detection/openvino_model.yaml
Comment (samet-akcay) [inline src/otx/recipe/detection/openvino_model.yaml:17]: If these are the only args that comes within `config`, I propose to explicitly define them in the base model signature

File: tests/integration/cli/test_cli.py
Comment (samet-akcay) [inline tests/integration/cli/test_cli.py]: Maybe something like this to see this reason in the logs too?

```python
pytest.skip("OMZ doesn't have proper model for Pytorch MaskRCNN interface. Need to change this test when export enabled")
```
ofc, we might need to divide the line, as it is a bit long :)

File: tests/unit/core/model/entity/test_base.py
Comment (samet-akcay) [inline tests/unit/core/model/entity/test_base.py]: how about?
```suggestion
    def batch(self) -> OTXBatchDataEntity:
```
 or `dataset_batch` or `input_batch`, but not `entity` :)

Comment (samet-akcay) [reply tests/unit/core/model/entity/test_base.py]: the reason is entity is not clear and too generic. In the future, I'll propose removing `Entity` from all of these objects :)

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Thanks @kprokofi! I have added some comments in the code base and also have a more general simple questions:
- How many OVModel is an OTX developer expected to maintain now. The ones in here and also ones in modelapi repo?
- If a developer is expected to maintain both of the openvino models, will this add extra maintenance effort?

Thanks again!

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py:230]: Actually, if model_name is not in config -> the error will be raised. I agree with you that we may refactor config parameters and maybe add some of them in constructor directly

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py:228]: Actually, num_classes was added by @vinnamkim in OTXModel for unifying representation of the number of classes. As we quickly discussed with Vlad above, we indeed could refactor it latter and add some parameters explicitly in constructor. However, it is not for that PR, I guess.
Let's address this issue a bit later

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py]: Ok

File: tests/unit/core/model/entity/test_base.py
Comment (kprokofi) [reply tests/unit/core/model/entity/test_base.py]: input_batch looks good for me

File: src/otx/recipe/detection/openvino_model.yaml
Comment (kprokofi) [reply src/otx/recipe/detection/openvino_model.yaml:17]: someone can define more parameters here. I will add them there.

File: src/otx/core/model/module/segmentation.py
Comment (kprokofi) [reply src/otx/core/model/module/segmentation.py:91]: We should prepare our outputs from the model in format suitable for metric computation, that is why we use self._convert_pred_entity_to_compute_metric(). Changing this logic requires more discussions, probably
I will rename `pred_list` to `predictions`

File: src/otx/core/model/entity/instance_segmentation.py
Comment (kprokofi) [reply src/otx/core/model/entity/instance_segmentation.py]: Ok

**Comment by kprokofi** [comment]
> Thanks @kprokofi! I have added some comments in the code base and also have a more general simple questions:
> 
> * How many OVModel is an OTX developer expected to maintain now. The ones in here and also ones in modelapi repo?
> * If a developer is expected to maintain both of the openvino models, will this add extra maintenance effort?
> 
> Thanks again!

I guess, we will maintain only our models come from OTX. Those models from OMZ should be maintained by according team. We only support its loading and inference via ModelAPI.
We will also add OVModel for multilabel and hierarhical classification. Also, OVMOdel for action task should be implemented, but there is a bit harder situation, since we don't have ModelAPI for Action Classification

File: src/otx/recipe/detection/openvino_model.yaml
Comment (kprokofi) [reply src/otx/recipe/detection/openvino_model.yaml:17]: image
Actually we have all required parameters here

File: tests/unit/core/model/entity/test_base.py
Comment (kprokofi) [reply tests/unit/core/model/entity/test_base.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/0c846b7d0e0e44feaf1f1e42bfd5202b122102a3)

File: tests/integration/cli/test_cli.py
Comment (kprokofi) [reply tests/integration/cli/test_cli.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/0c846b7d0e0e44feaf1f1e42bfd5202b122102a3)

File: src/otx/core/utils/build.py
Comment (kprokofi) [reply src/otx/core/utils/build.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/0c846b7d0e0e44feaf1f1e42bfd5202b122102a3)

File: src/otx/core/utils/build.py
Comment (kprokofi) [reply src/otx/core/utils/build.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/0c846b7d0e0e44feaf1f1e42bfd5202b122102a3)

File: src/otx/core/model/entity/instance_segmentation.py
Comment (kprokofi) [reply src/otx/core/model/entity/instance_segmentation.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/0c846b7d0e0e44feaf1f1e42bfd5202b122102a3)

File: src/otx/core/model/module/segmentation.py
Comment (kprokofi) [reply src/otx/core/model/module/segmentation.py:91]: [Done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/5fd6f80cf57fcaa7d8685169944bb8a28a801563)

File: src/otx/core/model/entity/classification.py
Comment (kprokofi) [reply src/otx/core/model/entity/classification.py:334]: [I added proper types](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/0c846b7d0e0e44feaf1f1e42bfd5202b122102a3)

File: src/otx/core/model/entity/base.py
Comment (kprokofi) [reply src/otx/core/model/entity/base.py]: [Done](https://github.com/openvinotoolkit/training_extensions/pull/2769/commits/0c846b7d0e0e44feaf1f1e42bfd5202b122102a3)

**Comment by kprokofi** [comment]
@samet-akcay , could you, please, check changes? It is important to merge it today/early tomorrow. We should start to check export - inference IR e2e results.

File: src/otx/core/model/entity/base.py
Comment (samet-akcay) [reply src/otx/core/model/entity/base.py:228]: alright. I see that `config` is passed to many other objects, so we could think about a wider refactor in the future addressing those too. Thanks!

File: src/otx/recipe/detection/openvino_model.yaml
Comment (samet-akcay) [reply src/otx/recipe/detection/openvino_model.yaml:17]: thanks, yeah, this is relevant to what we discussed above. We could address it later on then

**Review by samet-akcay — APPROVED** [review]
thanks for the effort!

**Comment by kprokofi** [comment]
@vinnamkim @jaegukhyun Could you approve once again and merge, please? Thanks!

**Review by jaegukhyun — APPROVED** [review]

**Review by vinnamkim — APPROVED** [review]

**Pull request merged**
